### PR TITLE
Add persistent window position, window maximization, and fullscreen state

### PR DESCRIPTION
**Pull request by freakode**
Implements the base functionality of #461.

Current Issues:
-Window position (0, 0) does not set the window to exactly the top left of the screen.  Bounds set to (-0.1, 1.1) in the configuration manager to work around this issue.

### CODE REVIEW

**Comment by peppy** [comment]
Weirdly, my window location doesn't seem to restore correctly consistently.

**Comment by peppy** [comment]
Looks like as you are only updating the config on invalidation, moving a window alone is not enough to save the position. May consider either binding to a LocationChanged event or (possibly the better option of?) only updating config values on dispose of BaseGame.

There's some formatting issues I have with your code, but let's get it in a working state first. Also please join us on discord if you are interested https://discord.gg/ppy

**Comment by freakode** [comment]
Interesting, I haven't had any issues with window position.  I'll look in to moving config save to the Dispose method.  Also I sent some messages in the Discord arrival.

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs [hunk 1]
```diff
@@ -22,13 +22,49 @@ public DesktopGameWindow() : base(default_width, default_height)
...
+                WindowState = value ? WindowState.Fullscreen : WindowState.Normal;
+            }
+
+            get
+            {
+                return WindowState == WindowState.Fullscreen;
+            }
+        }
+
+        public override bool Maximized
```
Comment (peppy) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: not sure we want to support "maximized". i think it should just enter borderless on becoming maximized.

**Comment by peppy** [comment]
Rather than having `bool fullscreen` `bool borderless`, might be easier to have
```
enum WindowMode
{
    Windowed,
    Borderless,
    Fullscreen
}
```

because you should never be fullscreen and not borderless, for instance.

**Comment by peppy** [comment]
This definitely seems to currently be breaking on switching to full-screen using alt-enter.

**Comment by freakode** [comment]
How exactly does it break?  Does anything draw at all after switching to full-screen?

**Comment by freakode** [comment]
What I changed from the last commit is actually using the values from the configuration file to set the full-screen resolution instead of automatically using the desktop resolution.  The only unexpected behavior that I am getting is that setting the resolution lower than my desktop's resolution puts the game in borderless fullscreen mode, which from earlier discussion seems like a bug in the framework.

**Comment by Tom94** [comment]
Using a lower resolution than the desktop resolution for fullscreen means we need to adjust the desktop's resolution first, before going into fullscreen. If OpenTK doesn't do this by itself, then the framework has to (or we extend OpenTK with such support).

**Comment by peppy** [comment]
I think all the window logic should be in BasicGameWindow, not BaseGame. Also not a huge fan of these `update` and `set` methods; it should hopefully all be contained in the get/set of individual properties.

File: osu.Framework/Configuration/FrameworkConfigManager.cs [hunk 2]
```diff
@@ -16,17 +16,19 @@ protected override void InitialiseDefaults()
...
             Set(FrameworkConfig.Height, 768, 480);
+            Set(FrameworkConfig.WindowedPositionX, 0.5, -0.1, 1.1);
+            Set(FrameworkConfig.WindowedPositionY, 0.5, -0.1, 1.1);
             Set(FrameworkConfig.AudioDevice, string.Empty);
             Set(FrameworkConfig.VolumeUniversal, 1.0, 0, 1);
             Set(FrameworkConfig.VolumeMusic, 1.0, 0, 1);
             Set(FrameworkConfig.VolumeEffect, 1.0, 0, 1);
-            Set(FrameworkConfig.HeightFullscreen, 9999, 240, 9999);
-            Set(FrameworkConfig.WidthFullscreen, 9999, 320, 9999);
+            Set(FrameworkConfig.WidthFullscreen, 1920, 320, 9999);
```
Comment (peppy) [inline osu.Framework/Configuration/FrameworkConfigManager.cs]: why this change? this breaks fullscreen on all my systems (my resolution is higher than the max you imposed in all cases).

**Comment by freakode** [comment]
I am now using these config values to set the full-screen resolution, whereas before it was using the current desktop resolution as the full-screen resolution.  When using the default of 9999x9999 a borderless full-screen window of that size is created.  I will look more in to proper resolution setting for full-screen to sort this whole thing out.

**Comment by peppy** [comment]
Also just to clarify, what I mean is passing a FrameworkConfigManager reference to BasicGameWindow (maybe in SetHost?) and on receiving it the gamewindow should read the defaults (and apply instantly) then handle updating the values itself.

**Comment by freakode** [comment]
Update: Figured how to do proper mode setting for full-screen mode.  Currently journeying through refactor-land to move the resolution changing logic to *GameWindow.

**Comment by freakode** [comment]
Still needs a bit of cleaning up, but it works now!

File: osu.Framework/Platform/BasicGameWindow.cs [hunk 3]
```diff
@@ -82,6 +79,11 @@ public void SetTitle(string title)
             Title = title;
         }
 
+        public virtual void SetupWindow(FrameworkConfigManager config)
+        {
+            Config = config;
```
Comment (peppy) [inline osu.Framework/Platform/BasicGameWindow.cs]: Rather than storing the config locally, it seems like a better fit will be to store all the bindables you need to local vairables using `GetBindable` (and not keep the config reference).

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs [hunk 4]
```diff
@@ -22,13 +24,111 @@ public DesktopGameWindow() : base(default_width, default_height)
...
+                {
+                    CurrentWindowMode = ((Bindable<WindowMode>)sender).Value;
+                };
+        }
+
+        public virtual void SaveWindow()
+        {
+            if (CurrentWindowMode == WindowMode.Fullscreen)
+            {
+                Config.Set(FrameworkConfig.WidthFullscreen, ClientSize.Width);
```
Comment (peppy) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: I'd probably allow the config to *not* be set, just to decouple this from having a config explicitly provided (make the incoming config argument in `SetupWindow` nullable and use `widthFullscreen?.Set(ClientSize.Width)`)

File: osu.Framework/Platform/BasicGameWindow.cs (hunk 3 above)
Comment (Tom94) [inline osu.Framework/Platform/BasicGameWindow.cs]: Might want to do it like AudioManager: use your own bindables and weld them to config bindables if a config is provided (or do it from outside; check how AudioManager is created).

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs (hunk 4 above)
Comment (Tom94) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: Applies also to here.

File: osu.Framework/Platform/BasicGameWindow.cs (hunk 3 above)
Comment (peppy) [inline osu.Framework/Platform/BasicGameWindow.cs]: that does avoid null checks and allow potential other sources to bind. good idea.

**Comment by peppy** [comment]
Apart from the welding bindable changes, there's one more thing:

Borderless should use the native resolution for the window. Currently it uses the last set window size, which can cause it to take up only a portion of the screen.

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs [hunk 5]
```diff
@@ -18,17 +20,132 @@ public class DesktopGameWindow : BasicGameWindow, IDropTarget
...
+            }
+        }
+
+        private void mode_ValueChanged(object sender, EventArgs e)
+        {
+            switch (mode.Value)
+            {
+                case WindowMode.Fullscreen:
+                    DisplayResolution newResolution = DisplayDevice.Default.SelectResolution(widthFullscreen, heightFullscreen, 0, 0);
+                    DisplayDevice.Default.ChangeResolution(newResolution);
```
Comment (Tom94) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: You might want to copy over the helper function we have in stable osu for changing resolution. There, when we try to go fullscreen from a non-default monitor it actually works on that other monitor. Also IIRC we have some decent fall-back logic in case changing resolution does not work.

Comment (peppy) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: yep, this is by no means complete, but it works perfectly for me (and i'd like to get a basic implementation merged to iterate on because i constantly have to merge this into my branch to make dev streamable at custom screen locations).

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs [hunk 6]
```diff
@@ -18,17 +20,132 @@ public class DesktopGameWindow : BasicGameWindow, IDropTarget
...
+                    WindowState = WindowState.Fullscreen;
+                    break;
+                case WindowMode.Borderless:
+                    DisplayDevice.Default.RestoreResolution();
+
+                    WindowState = WindowState.Maximized;
+                    WindowBorder = WindowBorder.Hidden;
+
+                    //must add 1 to enter borderless
+                    ClientSize = new Size(DisplayDevice.Default.Bounds.Width + 1, DisplayDevice.Default.Bounds.Height + 1);
```
Comment (Tom94) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: IIRC this was not sufficient on all hardware. I think we did some more tricks than just this on stable to get everything right. Did you double-check?

Comment (peppy) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: this is to get *non*-exclusive full screen, and is very much a hack for now.

Comment (peppy) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: ie. the opposite of what you'd expect to not work

File: osu.Framework/BaseGame.cs [hunk 7]
```diff
@@ -113,7 +108,8 @@ public virtual void SetHost(BasicGameHost host)
                 Config = new FrameworkConfigManager(host.Storage);
 
             this.host = host;
-            host.Size = new Vector2(Config.Get<int>(FrameworkConfig.Width), Config.Get<int>(FrameworkConfig.Height));
+
+            host.Window?.SetupWindow(Config);
```
Comment (Tom94) [inline osu.Framework/BaseGame.cs]: Alternatively, can make setting the title of the window part of SetupWindow.

File: osu.Framework/BaseGame.cs [hunk 8]
```diff
@@ -237,6 +233,12 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
                 }
             }
 
+            if (state.Keyboard.AltPressed && args.Key == Key.Enter)
+            {
+                Window.CycleMode();
```
Comment (Tom94) [inline osu.Framework/BaseGame.cs]: We might want to allow headless to receive input events, or at least to have them manually triggered for testing purposes, so `Window?.CycleMode()`

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs (hunk 6 above)
Comment (Tom94) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: I'm aware of what this is for. :p 
Pretty sure it failed to leave exclusive fullscreen on some people I tested with. I remember debugging this stuff for many hours with people from the tech support forums.

File: osu.Framework.Desktop/Platform/DesktopGameWindow.cs [hunk 9]
```diff
@@ -36,8 +36,6 @@ public DesktopGameWindow() : base(default_width, default_height)
 
         public override void SetupWindow(FrameworkConfigManager config)
```
Comment (Tom94) [inline osu.Framework.Desktop/Platform/DesktopGameWindow.cs]: Is there any more reason to even have the `config` argument?
